File: backend/api/src/main/java/com/stellarcompact/api/faction/FactionConfigService.java
package com.stellarcompact.api.faction;

import com.stellarcompact.agentruntime.prompt.AssembledPrompt;

/**
 * The application-service seam behind the Sovereign-config REST endpoints (board card
 * E6-02; rest-api spec, section Sovereign configuration). It owns the per-seat
 * configuration registry, the owner binding, and the lifecycle gate on directive edits.
 *
 * <p><b>Why an interface (the E6-01/E6-04 seam pattern).</b> Like {@code MatchService}
 * and {@code FactionOwnershipRegistry}, configuration storage is behind a seam so the
 * in-memory backing ({@link InMemoryFactionConfigService}) can be swapped for a
 * persistence-backed one (a {@code faction_config} table joined to the seat-assignment
 * table) without touching the controller or its DTOs.
 *
 * <p><b>Owner is a parameter, resolved server-side.</b> Every method that depends on
 * owner identity takes the resolved principal name as an explicit argument; the
 * controller derives it from the authenticated principal / owner token, never from a
 * client-settable body field (principle 2). The service treats a {@code null}/blank
 * principal as "not the owner" (default deny for writes; redaction for reads).
 *
 * <p><b>This config feeds prompt assembly.</b> {@link #assembledPromptFor} renders a
 * seat's stored {@code SovereignConfig} through the agent-runtime {@code PromptAssembler}
 * (E4-02) so the prompt a configured seat would receive reflects the persona/goals/
 * constraints stored here - the direct, testable link to the LLM path (orchestrator
 * {@code LlmSeat}).
 */
public interface FactionConfigService {

    /**
     * Create/attach a Sovereign configuration to a seat in {@code gameId} and bind
     * {@code ownerPrincipal} as its owner. The seat is chosen as the first unconfigured
     * seat in the match (ascending id order). Idempotent re-attach by the same owner
     * updates the existing config; a different principal attaching to an already-owned
     * seat is rejected.
     *
     * @param gameId         the match
     * @param ownerPrincipal the resolved owning principal (non-blank)
     * @param request        the seat configuration
     * @return the attached seat handle
     * @throws com.stellarcompact.api.match.MatchNotFoundException if no such match
     * @throws NotFactionOwnerException if {@code ownerPrincipal} is blank, or no free seat
     */
    AttachFactionResponse attach(String gameId, String ownerPrincipal, CreateFactionRequest request);

    /**
     * Read a seat's configuration. The view is the full owner view iff
     * {@code requesterPrincipal} is the seat's registered owner; otherwise it is the
     * redacted public view (identity only). A non-owner read is NOT an error.
     *
     * @throws FactionNotFoundException if no such faction handle
     */
    FactionConfigView view(String factionHandle, String requesterPrincipal);

    /**
     * Apply a standing-directive edit. Owner-gated (non-owner -&gt; 403) and lifecycle-gated
     * (match RUNNING -&gt; 409); editable only between matches.
     *
     * @throws FactionNotFoundException   if no such faction handle
     * @throws NotFactionOwnerException   if the requester is not the owner
     * @throws DirectivesLockedException  if the owning match is RUNNING
     */
    FactionConfigView updateDirectives(
            String factionHandle, String requesterPrincipal, UpdateDirectivesRequest request);

    /**
     * Assemble the system+user prompt a configured seat would receive, from its stored
     * {@code SovereignConfig} (the link from this config to prompt assembly, E4-02). The
     * {@code worldView} is any token-compact JSON-serialisable perception object (the
     * orchestrator passes its own {@code WorldView}); tests pass a tiny stand-in.
     *
     * @throws FactionNotFoundException if no such faction handle
     */
    AssembledPrompt assembledPromptFor(String factionHandle, Object worldView);
}

File: frontend/src/app/services/faction-config.service.ts
import { Injectable, InjectionToken, inject } from '@angular/core';
import {
  HttpClient,
  HttpHeaders,
  HttpErrorResponse,
} from '@angular/common/http';
import { firstValueFrom } from 'rxjs';

// ---------------------------------------------------------------------------
// Dev owner-token injection token — stand-in until session auth lands.
// Override in tests or app config with a real token value.
// ---------------------------------------------------------------------------

export const OWNER_TOKEN = new InjectionToken<string>('OWNER_TOKEN', {
  providedIn: 'root',
  factory: () => '',
});

// ---------------------------------------------------------------------------
// Contract types — mirror the E6-02 backend DTOs exactly
// ---------------------------------------------------------------------------

/** Model tier enum — orchestration routing only, never a vendor name. */
export type ModelTier = 'SMALL' | 'MEDIUM' | 'LARGE';

/** Body for POST /api/games/{gameId}/factions */
export interface CreateFactionRequest {
  /** One of personaPreset or customPersona must be provided. */
  personaPreset?: string;
  customPersona?: string;
  goals: string[];
  hardConstraints: string[];
  modelTier: ModelTier;
}

/** Response from POST /api/games/{gameId}/factions */
export interface AttachFactionResponse {
  factionId: string;
  gameId: string;
  seatId: string;
}

/** Full / redacted view from GET /api/factions/{factionId} */
export interface FactionConfigView {
  factionId: string;
  gameId: string;
  seatId: string;
  configured: boolean;
  /** true when the requester is the registered owner. */
  owner: boolean;
  /** null when redacted */
  persona: string | null;
  goals: string[];
  hardConstraints: string[];
  /** null when redacted */
  modelTier: ModelTier | null;
}

/** Body for PATCH /api/factions/{factionId} */
export interface UpdateDirectivesRequest {
  goals?: string[];
  hardConstraints?: string[];
  modelTier?: ModelTier;
  customPersona?: string;
}

/** Structured error for the between-matches lifecycle gate (409). */
export class DirectivesLockedException extends Error {
  constructor() {
    super('Faction directives are locked while the match is RUNNING.');
    this.name = 'DirectivesLockedException';
  }
}

// ---------------------------------------------------------------------------
// Service
// ---------------------------------------------------------------------------

/**
 * Faction config service — thin HTTP wrapper around the E6-02 faction config
 * REST surface.
 *
 *   POST /api/games/{gameId}/factions   → attach/create Sovereign config
 *   GET  /api/factions/{factionId}      → read config view (owner-sensitive)
 *   PATCH /api/factions/{factionId}     → update standing directives
 *
 * The `X-Owner-Token` dev header is injected from the OWNER_TOKEN token so it
 * can be overridden per-environment without code changes (principle 4).
 * Owner identity is resolved server-side from this header (or real auth later).
 */
@Injectable({ providedIn: 'root' })
export class FactionConfigService {
  private readonly http = inject(HttpClient);
  private readonly ownerToken = inject(OWNER_TOKEN);

  private get headers(): HttpHeaders {
    let h = new HttpHeaders({ 'Content-Type': 'application/json' });
    if (this.ownerToken) {
      h = h.set('X-Owner-Token', this.ownerToken);
    }
    return h;
  }

  /**
   * Attach a Sovereign config to the first free seat in the given game.
   * Returns 201 AttachFactionResponse on success.
   */
  async attach(
    gameId: string,
    body: CreateFactionRequest,
  ): Promise<AttachFactionResponse> {
    return firstValueFrom(
      this.http.post<AttachFactionResponse>(
        `/api/games/${gameId}/factions`,
        body,
        { headers: this.headers },
      ),
    );
  }

  /**
   * Read the faction config view for a given factionId.
   * Returns the full view when the requester is the owner; otherwise returns
   * the public identity only (persona/goals/constraints/tier are redacted).
   */
  async getConfig(factionId: string): Promise<FactionConfigView> {
    return firstValueFrom(
      this.http.get<FactionConfigView>(`/api/factions/${factionId}`, {
        headers: this.headers,
      }),
    );
  }

  /**
   * Update standing directives for a faction (between matches only).
   * Throws DirectivesLockedException on 409 (match is RUNNING).
   * Throws on 403 when the requester is not the owner.
   */
  async updateDirectives(
    factionId: string,
    body: UpdateDirectivesRequest,
  ): Promise<FactionConfigView> {
    try {
      return await firstValueFrom(
        this.http.patch<FactionConfigView>(`/api/factions/${factionId}`, body, {
          headers: this.headers,
        }),
      );
    } catch (err) {
      if (err instanceof HttpErrorResponse && err.status === 409) {
        throw new DirectivesLockedException();
      }
      throw err;
    }
  }
}
